src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -549,29 +549,60 @@ public Stream GetStream(int len, bool canSeek)
 
         /// <summary>
         /// Seeks the first null terminator (\0) and returns the string up to it. The buffer must already
         /// contain the entire string and its terminator. If any character could not be decoded, a question
         /// mark character is returned instead of throwing an exception.
         /// </summary>
-        public string ReadNullTerminatedStringRelaxed() => ReadNullTerminatedString(RelaxedTextEncoding);
+        public string ReadNullTerminatedStringRelaxed()
+            => ReadNullTerminatedString(RelaxedTextEncoding, async: false).GetAwaiter().GetResult();
+
+        public ValueTask<string> ReadNullTerminatedString(bool async, CancellationToken cancellationToken = default)
+            => ReadNullTerminatedString(TextEncoding, async, cancellationToken);
 
         /// <summary>
-        /// Seeks the first null terminator (\0) and returns the string up to it. The buffer must already
-        /// contain the entire string and its terminator.
+        /// Seeks the first null terminator (\0) and returns the string up to it. Reads additional data from the network if a null
+        /// terminator isn't found in the buffered data.
         /// </summary>
-        /// <param name="encoding">Decodes the messages with this encoding.</param>
-        string ReadNullTerminatedString(Encoding encoding)
+        ValueTask<string> ReadNullTerminatedString(Encoding encoding, bool async, CancellationToken cancellationToken = default)
         {
-            int i;
-            for (i = ReadPosition; Buffer[i] != 0; i++)
-                Debug.Assert(i <= ReadPosition + ReadBytesLeft);
-            Debug.Assert(i >= ReadPosition);
-            var result = encoding.GetString(Buffer, ReadPosition, i - ReadPosition);
-            ReadPosition = i + 1;
-            return result;
+            return ReadFromBuffer(out var s)
+                ? new ValueTask<string>(s)
+                : ReadLong(s);
+
+            bool ReadFromBuffer(out string s)
+            {
+                var start = ReadPosition;
+                while (ReadPosition < FilledBytes)
+                {
+                    if (Buffer[ReadPosition++] == 0)
+                    {
+                        s = encoding.GetString(Buffer, start, ReadPosition - start - 1);
+                        return true;
+                    }
+                }
+
+                s = encoding.GetString(Buffer, start, ReadPosition - start);
+                return false;
+            }
+
+            async ValueTask<string> ReadLong(string s)
+            {
+                var builder = new StringBuilder(s);
+                bool complete;
+                do
+                {
+                    await ReadMore(async, cancellationToken);
+                    complete = ReadFromBuffer(out s);
+                    builder.Append(s);
+                }
+                while (!complete);
+
+                return builder.ToString();
+            }
```

Comment:
Maybe a bit late, but I think this code will fail if the split is in the middle of a unicode character that uses more than 1 byte to represent. Better `Ensure` the max length of a string is already present in the buffer (as it was previously), or create a temporary byte buffer with all the data which you then convert to a string in one go using the `encoding`.

Reply:
You're technically right... However, note that this is for reading null-terminated strings, which AFAIK are only used for certain short ASCII control strings, rather than actual user data - so I doubt there's a real bug here. But you can open a new issue to at least have it tracked.

BTW if we do want to preemptively fix this, then rather than entirely buffering the strings them in a temporary buffer, we can use [System.Text.Decoder](https://docs.microsoft.com/en-us/dotnet/api/system.text.decoder?view=net-5.0) to incrementally load more chunks a decode. Note that we haven't even done this optimization for actual user strings (which are probably much more important, see TextHandler.ReadLong).

Reply:
At least in the replication protocol they are used for user data as well, for example table names, which can contain non-ascii characters.

I don't see why System.Text.Decoder would be anything better. It gives you some `char`s on every iteration. These chars are temporary anyway and must then be made into a string, after which the previously buffered chars are discarded. But generally a char[] takes up more memory than the corresponding byte[] in UTF8, so I think it's better to just buffer everything as bytes. The encoder can then convert these to a string in-place, with no extra temporary space.

Reply:
You're right... IIRC that's why I didn't do it in TextHandler either.

Open an issue for us to fix this?